### PR TITLE
[pytorch] handle pybind11 style registration API with code analyzer

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36607 [pytorch] handle pybind11 style registration API with code analyzer**
* #35941 [pytorch] improve code analyzer to dump ops called from c++ functions

Summary:
PR #36258 and subsequent PRs in the stack switch c10 registrations to
the new pybind11 style registration API. One notable difference from old
c10 registration API is that, operator's namespace is no longer in op
schema string, e.g. "aten::" will be factored out from "aten::conv",
"aten::emtpy" and etc. The namespace string will be declared at the
beginning of registrations with TORCH_LIBRARY / TORCH_LIBRARY_IMPL
macro.

A rather simple fix is to extract namespace string from the name of
enclosing function of registrations, as the TORCH_LIBRARY macro will
always create an init function (per namespace) by appending namespace
string to a common prefix.

Another side effect of the API change is that it adds some debug string
constants to the registration API, and because of factoring out the
namespace part from op name, there is no longer an effect way to
differentiate between real op name and debug strings. A simple
workaround is that we only keep the first string constant it encounters
while BFSing the LLVM IR - the real op name is directly passed into the
registration call while the debug string is indirectly passed via
CppFunction.

These new assumptions might be broken by future changes but it's so simple
to implement to unblock the API work.

Differential Revision: [D21026008](https://our.internmc.facebook.com/intern/diff/D21026008)

### CODE REVIEW

File: tools/code_analyzer/op_dependency.cpp [hunk 1]
```diff
@@ -510,16 +533,32 @@ class OpDependency : public ModulePass {
...
+      if (auto schemaStr = extractOpSchema(contextualNamespace, V)) {
         if (visitedOps) {
-          (*visitedOps).insert(*schemaStr);
+          // NB: Some debug string constants might be connected to the
+          // registration instruction, e.g.: "Lambda". Since we has factored
+          // out namespace from op schema string, there is no longer a simple
+          // way to identify these fake ops. For now we simply take the first
+          // instance as the actual op name is cloest to the seed instruction
+          // in BFS order.
+          if (!visitedOps->empty()) {
```
Comment (ljk53) [inline tools/code_analyzer/op_dependency.cpp:552]: We print warning message when `size(ops) != 1` per registration/invocation - in reality it never found >1 ops in past months until now. So I decided to take the simple workaround for now.

File: tools/code_analyzer/op_dependency.cpp [hunk 2]
```diff
@@ -149,6 +163,15 @@ cl::opt<RegexOpt, true, cl::parser<std::string>> RootSymbolPattern(
              "Example: -root_symbol_pattern 'torch::jit'"),
     cl::location(RootSymbolPatternLoc));
 
+cl::list<RegexOpt, bool, RegexOptParser> TorchLibraryInitPattern(
+    "torch_library_init_pattern",
+    cl::desc("Regular expression used to identify TorchLibraryInit symbols "
+             "that are generated by `TORCH_LIBRARY` macro. The first capturing "
+             "group is used to extract namespace string. "
```
Comment (ezyang) [inline tools/code_analyzer/op_dependency.cpp:170]: Haha, clever!

File: tools/code_analyzer/run_analyzer.sh [hunk 3]
```diff
@@ -14,10 +14,12 @@ echo "Analyze: ${INPUT}"
...
 # to operate, so for safety we match a more expansive set.
 "${ANALYZER_BIN}" \
-  -op_schema_pattern="^(_aten|_prim|aten|quantized|profiler|_test)::[^ ]+" \
-  -op_register_pattern="c10::RegisterOperators::(op|checkSchemaAndRegisterOp_)|c10::Module::(_?def|_?impl|impl_UNBOXED)" \
+  -op_schema_pattern="^(_aten|_prim|aten|quantized|profiler|_test)::[a-zA-Z0-9_.]+(\(.*)?$" \
+  -op_register_pattern="c10::RegisterOperators::(op|checkSchemaAndRegisterOp_)|c10::Module::(_?def|_?impl|impl_UNBOXED)|c10::Library::(_?def|_?impl|_?impl_UNBOXED)" \
   -op_invoke_pattern="c10::Dispatcher::findSchema|callOp" \
   -root_symbol_pattern="torch::jit::[^(]" \
+  -torch_library_init_pattern="^.*TORCH_LIBRARY_init_([^(]+)(\(.*)?$" \
+  -torch_library_init_pattern="^.*TORCH_LIBRARY_IMPL_init_([^(]+)_([^_]+)(\(.*)?$" \
```
Comment (ezyang) [inline tools/code_analyzer/run_analyzer.sh:22]: All hail the glory of backtracking. But I guess we will have to be careful to never define a Backend with an underscore in it...

File: tools/code_analyzer/op_dependency.cpp [hunk 4]
```diff
@@ -510,16 +533,32 @@ class OpDependency : public ModulePass {
...
       SET* visitedOps, SET* visitedFunctions) {
     std::shared_ptr<VALUE_MAP> debugPath =
         (Verbose > 2 ? std::make_shared<VALUE_MAP>() : nullptr);
     auto callback = [&](Value* V) -> void {
-      if (auto schemaStr = extractOpSchema(V)) {
+      if (auto schemaStr = extractOpSchema(contextualNamespace, V)) {
         if (visitedOps) {
-          (*visitedOps).insert(*schemaStr);
+          // NB: Some debug string constants might be connected to the
+          // registration instruction, e.g.: "Lambda". Since we have factored
```
Comment (ezyang) [inline tools/code_analyzer/op_dependency.cpp:547]: So, I can add a prefix to the "debug" strings, they're totally unstructured and I can put whatever I want in them

File: tools/code_analyzer/op_dependency.cpp [hunk 5]
```diff
@@ -710,11 +768,14 @@ class OpDependency : public ModulePass {
     }
   }
 
-  static std::shared_ptr<std::string> extractOpSchema(Value* V) {
+  static std::shared_ptr<std::string> extractOpSchema(
+      const std::string& contextualNamespace, Value* V) {
```
Comment (ezyang) [inline tools/code_analyzer/op_dependency.cpp:772]: note to self: `contextualNamespace` comes with the trailing `::` attached already

File: tools/code_analyzer/op_dependency.cpp [hunk 6]
```diff
@@ -710,11 +768,14 @@ class OpDependency : public ModulePass {
...
     std::vector<std::string> schemaStrs;
     extractStringValue(V, [&](const std::string& str) {
-      if (FunctionSchemaPatternLoc.pattern->match(str)) {
-        schemaStrs.push_back(str);
+      const std::string& schemaStr =
```
Comment (ezyang) [inline tools/code_analyzer/op_dependency.cpp:775]: I love the smell of const reference extension in the morning

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
How do you want to coordinate landing this? Should I just commandeer this into my patchset?

File: tools/code_analyzer/run_analyzer.sh (hunk 3 above)
Comment (ljk53) [reply tools/code_analyzer/run_analyzer.sh:22]: We can put some special delimiter when concatenating the init symbol name in the macro and update the RE.

File: tools/code_analyzer/op_dependency.cpp (hunk 6 above)
Comment (ljk53) [reply tools/code_analyzer/op_dependency.cpp:775]: yeah, it's intentional to avoid copying in no-namespace-case - is it bad? :>

**Pull request closed without merging**